doc/cli_subprocess_guidelines.md

```diff
@@ -0,0 +1,110 @@
+3) it might be not that obvious to find target command's executable file **in windows platform**, a tool developer can use is `shutil.which` that gives the executable file path in windows system, like `shutil.which(git)`. The cmd `git --version` can be adjusted as `[shutil.which(git), "--version"]`. Please provide the corresponding executable path in target platforms.
+4) if the target cmd is az-related, like `az group show --name xxxx`, please use internal corresponding func call to get the target information.
+
+
+### Best Practices In Subprocess Use Cases
+
+
+The following sections discuss some secure coding conventions that, when implemented, can help protect cli applications from command injection vulnerabilities when calling subsystems through `subprocess`.
+
+#### Proper input validation and sanitization
+
+Input from users or external sources should never be trusted when appending them into subsystem's cmd. Developers better provide expected patterns to validate and sanitize user input before adding that into subsystem's cmd. 
+Below is an example input sanitizer that only allows alphanumeric characters from the input string.  
+
+```commandline
+import re
+def sanitize_input(user_input):
+    return re.sub(r'[^a-zA-Z0-9]', '', user_input)
+```
+In a real cli module scenario, developers better use their corresponding sanitization pattern for reducing the chaining commands risks.
+
+#### Use parameterized cmd statement
+
+A parameterized cmd statement is a way to structure a command starting from defining the command structure first and then provide the parameters that should be inserted into the command separately. 
+In this way, the system will treat the first item as the executable file and the left item as args of that executable application.
+```
+# instead of
+cmd = f"git {user_input}"
+# using 
+cmd = ["git", user_input]
+```
+
+#### Avoid use `shell=True` with user_inputs
+
+When using subprocess module, avoid `shell=True` argument when it comes with cmd containing user inputs. When python executes a command with `shell=True`, the system shell will interpret any string passed in and if this string contains user input without proper sanitization, an attacker can inject malicious commands, as demonstrated in the very beginning of this doc.
+
+
+## Summary
+Ensuring the safety of cli from command injection under subprocess calling requires an in-depth understanding of these vulnerabilities and also proactive measures to counteract potential exploits. Cli developers can either apply the three security practices, if applicable, when using builtin `subprocess`, or use a more centralized func `run_cmd` cli provided, to safeguard cli modules from command injection attack and for future more accessible security enforcements.
```

Comment:
Avoid to use abbreviation `func`, use `function` instead which is more straightforward.